common/djangoapps/entitlements/api/v1/views.py

```diff
@@ -30,6 +32,34 @@ def get_queryset(self):
             return CourseEntitlement.objects.all().select_related('user')
         return CourseEntitlement.objects.filter(user=user).select_related('user')
 
+    def retrieve(self, request, *args, **kwargs):
+        """
+        Override the retrieve method to expire a record that is past the
+        policy and is requested via the API before returning that record.
+        """
+        entitlement = self.get_object()
+        entitlement.update_expired_at()
+        serializer = self.get_serializer(entitlement)
+        return Response(serializer.data)
+
+    def list(self, request, *args, **kwargs):
+        """
+        Override the list method to expire records that are past the
+        policy and requested via the API before returning those records.
+        """
+        queryset = self.filter_queryset(self.get_queryset())
+        with transaction.atomic():
+            for entitlement in queryset:
```

Comment:
This will result in a table scan and insert for every expired entitlement in the table when staff users make requests to this endpoint.